File: Algorithm/sortLinkList/quicksort.cpp
#include <iostream>
#include <vector>
using namespace std;
struct ListNode
{
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

void quicksort(vector<int> &arr, int low, int high)
{
  if (low < high)
  {
    int middle = partition(arr, low, high);
    quicksort(arr, low, middle - 1);
    quicksort(arr, middle + 1, high);
  }
}

class Solution
{
public:
  ListNode *quickSortList(ListNode *head)
  {
    // IMPORTANT: Please reset any member data you declared, as
    // the same Solution instance will be reused for each test case.
    //链表快速排序
    if (head == NULL || head->next == NULL)
      return head;
    qsortList(head, NULL);
    return head;
  }
  void qsortList(ListNode *head, ListNode *tail)
  {
    //链表范围是[low, high)
    if (head != tail && head->next != tail)
    {
      ListNode *mid = partitionList(head, tail);
      qsortList(head, mid);
      qsortList(mid->next, tail);
    }
  }
  ListNode *partitionList(ListNode *low, ListNode *high)
  {
    //链表范围是[low, high)
    int key = low->val;
    ListNode *loc = low;
    for (ListNode *i = low->next; i != high; i = i->next)
      if (i->val < key)
      {
        loc = loc->next;
        swap(i->val, loc->val);
      }
    swap(loc->val, low->val);
    return loc;
  }
};

File: Algorithm/sortLinkList/quickSort.js
function quickSort(arr, left, right) {
  const len = arr.length;
  let partitionIndex;
  left = typeof left != 'number' ? 0 : left;
  right = typeof right != 'number' ? len - 1 : right;

  if (left < right) {
    partitionIndex = partition(arr, left, right);
    quickSort(arr, left, partitionIndex - 1);
    quickSort(arr, partitionIndex + 1, right);
  }
  return arr;
}

function partition(arr, left, right) {     // 分区操作
  const pivot = left;                      // 设定基准值（pivot）
  let index = pivot + 1;
  for (let i = index;i <= right;i++) {
    if (arr[i] < arr[pivot]) {
      swap(arr, i, index);
      index++;
    }
  }
  swap(arr, pivot, index - 1);
  return index - 1;
}

function swap(arr, i, j) {
  const temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}


let waitingArr = [4, 5, 64, 4, 54, 56, 59, 64, 54, 4, 4, 45, 4]

console.log(quickSort(waitingArr,1,5))